veg2hab/vegkartering.py

```diff
@@ -1157,7 +1336,138 @@ def functionele_samenhang(self) -> pd.DataFrame:
+        )
+        rename_private_cols = {
+            "VegTypeInfo": "_VegTypeInfo",
+            "HabitatVoorstel": "_HabitatVoorstel",
+            "HabitatKeuze": "_HabitatKeuze",
+        }
+        gdf = self.gdf.rename(columns=rename_private_cols)
+
+        gdf = pd.concat([gdf, habkeuzes_df], axis=1)
+
+        gdf["_VegTypeInfo"] = (
+            gdf["_VegTypeInfo"].apply(VegTypeInfo.serialize_list).astype("string")
+        )
+        gdf["_HabitatKeuze"] = (
+            gdf["_HabitatKeuze"].apply(HabitatKeuze.serialize_list).astype("string")
+        )
+        gdf["_HabitatVoorstel"] = (
+            gdf["_HabitatVoorstel"]
+            .apply(HabitatVoorstel.serialize_list2)
+            .astype("string")
+        )
+
+        column_order = [
+            *self.PREFIX_COLS,
+            *habkeuzes_df.columns,
+            "_HabitatKeuze",
+            "_HabitatVoorstel",
+            "_VegTypeInfo",
+            *self.POSTFIX_COLS,
+        ]
+        # assert set(gdf.columns) - set(column_order)
+
+        gdf = gdf[column_order]
+
+        # annoying ARCGIS
+        rename_columns = {
+            col: "INTERN" + col for col in column_order if col.startswith("_")
+        }
+        gdf = gdf.rename(columns=rename_columns)
+
+        return gdf
+
+    @staticmethod
+    def _multi_col_to_habkeuze(row: pd.Series) -> List[Tuple[str, str, str]]:
+        result = []
+        for idx in range(1, 100):  # arbitrary number
+            habtype = row.get(f"Habtype{idx}", None)
+            habkeuze = row.get(f"Kwal{idx}", None)
+            opm = row.get(f"Opm{idx}", None)
+            if habtype is None and habkeuze is None:
+                break
+            result.append((habtype, habkeuze, opm))
+        else:
+            raise ValueError("Er zijn te veel kolommen met Habtype/Kwal")
+
+        return result
+
+    @classmethod
+    def from_editable_habtypes(cls, gdf: gpd.GeoDataFrame) -> Self:
+        # rename the INTERN columns
+        rename_columns = {
+            col: col[len("INTERN") :]
+            for col in gdf.columns
+            if col.startswith("INTERN_")
+        }
+        gdf = gdf.rename(columns=rename_columns)
+
+        # unpack json strings
+        for col, deserialization_func in {
+            "_VegTypeInfo": VegTypeInfo.deserialize_list,
+            "_HabitatKeuze": HabitatKeuze.deserialize_list,
+            "_HabitatVoorstel": HabitatVoorstel.deserialize_list2,
+        }.items():
+            gdf[col] = gdf[col].apply(deserialization_func)
+
+        # check for changed habitatkeuzes
+        altered_habkeuzes = gdf.apply(cls._multi_col_to_habkeuze, axis=1)
+        for row_idx, (new_keuzes, old_keuzes) in enumerate(
+            zip(altered_habkeuzes, gdf["_HabitatKeuze"])
+        ):
+            if len(new_keuzes) != len(old_keuzes):
+                logging.error(
+                    "Het aantal habitatkeuzes is veranderd. Wij kunnen niet garanderen dat de output correct is."
```

Comment:
Misschien habitatkeuzes hier veranderen naar complexdelen? Dan weet de ecoloog ook wat we bedoelen.